### PR TITLE
Single point-of-entry for deserialization

### PR DESCRIPTION
**Pull request by AdamGleave**
Saves the class of the object, so that `Serializable.load` can always de-serialize. This significantly simplifies `rewards/serialize.py` as no longer need to keep track of different reward types. (I'm still keeping the reward registry since not *everything* is serialized using our method, e.g. hard-coded reward functions or reward models from other codebases.)

This does feel an awful lot like re-inventing `pickle`, however trying to push everything into `pickle` would (a) not play nice with `tf.Checkpoint` and (b) having a hierarchical directory structure for our serialized models is nice and `pickle` can't do this.

Still not quite satisfied with how we're serializing/deserializing things so I'd welcome suggestions on improvements.

### CODE REVIEW

**Comment by AdamGleave** [comment]
@shwang would appreciate it if you could review this today or tomorrow, as some of my internal code depends on this change.

**Comment by shwang** [comment]
Ok will review today

File: src/imitation/util/serialize.py [hunk 1]
```diff
@@ -20,10 +20,20 @@ class Serializable(ABC):
   @abstractmethod
   def load(cls, directory):
     """Load object plus weights from directory."""
+    with open(os.path.join(directory, 'loader'), 'rb') as f:
+      load_cls = pickle.load(f)
+    return load_cls.load(directory)
```
[the remark this answers is not in the document]
Comment (shwang) [reply src/imitation/util/serialize.py]: The `Serializable.load` path ends up calling the actual implementation of `DiscrimNet.load`. Maybe there's something I don't understand about how name resolution works for `DiscrimNet.load`, but I thought that `DiscrimNet.load` should lead to the `def load(cls, ...)` defined in `DiscrimNet`.

File: src/imitation/util/serialize.py [hunk 2]
```diff
@@ -20,10 +20,20 @@ class Serializable(ABC):
...
   @abstractmethod
   def save(self, directory):
     """Save object and weights to directory."""
+    os.makedirs(directory, exist_ok=True)
+
+    load_path = os.path.join(directory, 'loader')
+    with open(load_path + '.tmp', 'wb') as f:
+      pickle.dump(type(self), f)
+    # Ensure atomic write
+    os.replace(load_path + '.tmp', load_path)
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/util/serialize.py]: This was to support the use case of calling `save` repeatedly with the same directory. `tf.train.Checkpoint` explicitly supports this use case: repeated calls to `save` just add a new checkpoint file. If the code got interrupted in the middle of saving a checkpoint, I don't want it to corrupt previously saved checkpoints in that directory.

Having said that, we might want to just not support this use case, since there's no guarantee that implementations other than `LayersSerializable` would support this.

File: src/imitation/rewards/serialize.py [hunk 3]
```diff
@@ -13,56 +13,34 @@
 RewardLoaderFn = Callable[[str, VecEnv], ContextManager[RewardFn]]
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/rewards/serialize.py]: OK, I've renamed it.

File: src/imitation/rewards/serialize.py [hunk 4]
```diff
@@ -13,56 +13,34 @@
...
 RewardNetLoaderFn = Callable[[str, VecEnv], reward_net.RewardNet]
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/rewards/serialize.py]: done.

File: src/imitation/util/serialize.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/util/serialize.py]: This is working as intended, but on reflection I think the design may be too confusing.

The desiderata here are:
  - I should be able to load any `Serializable` object via a single method without worrying about what type it is.
  - Different type objects should be able to support arbitrary serialization/deserialization logic, due to the heterogenity of things we might want to save. (For example, I have some reward models that wrap other reward models -- so serializing them means first serializing the inner model, then the outer model.)

I implemented this by:
  - Making `Serializable.save` save the type of the object into `loader`. Concrete implementations of this call it via super().
  - Concrete implementations of `Serializable` *must* override `load`, since it's abstract. (Otherwise `Serializable.load` could end up calling itself, which would be bad.)
  - `Serializable.load` read from `loader` and calls the type-specific `load`. This works since any concrete type must have its own `load`.

The confusion here is that `DiscrimNet` is itself an abstract class, and there's no `DiscrimNet.load`, since `DiscrimNetAIRL` and `DiscrimNetGAIL` do quite different things. So in this case `DiscrimNet.load` just goes through to `Serializable.load`, which then calls `DiscrimNetAIRL.load` or `DiscrimNetGAIL.load` as appropriate.

I'd appreciate alternative design suggestions: calling an abstract method, while legal Python, does feel odd. One alternative would be to just add a method `load_serialize` which does what `Serializable.load` currently does. I do like that e.g. `DiscrimNet.load` self-documents that we expect it to be a `DiscrimNet` (and potentially enforces this via type-checking), which would be lost with a generic helper method.

[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/util/serialize.py]: Note this is also true for `reward_net.RewardNet.load`, as there is no `load` method implemented in `RewardNet` (also abstract).

[the remark this answers is not in the document]
Comment (shwang) [reply src/imitation/util/serialize.py]: Thanks for writing up such a thorough explanation!

Yesterday when I was reading through the code I mistakenly confused the `DiscrimNetAIRL.load` for a `DiscrimNet.load` implementation. So I thought that `DiscrimNet.load` had an implementation, which was somehow being skipped in favor of `Serializable.load`. (Good to know now that my understanding of Python name resolution isn't broken).

**Comment by shwang** [comment]
I noticed that we no longer use `build_loader_fn_requires_path` from codecov. Consider deleting

```
(env) shwangster@airbears2-10-142-34-171:~/repos/imitation$ ag build_loader_fn_require_path
src/imitation/util/registry.py
92:def build_loader_fn_require_path(fn: Callable[[str], T],
```

**Comment by AdamGleave** [comment]
> I noticed that we no longer use `build_loader_fn_requires_path` from codecov. Consider deleting
> 
> ```
> (env) shwangster@airbears2-10-142-34-171:~/repos/imitation$ ag build_loader_fn_require_path
> src/imitation/util/registry.py
> 92:def build_loader_fn_require_path(fn: Callable[[str], T],
> ```

Thanks, done.
